ext/openssl/openssl.c

```diff
@@ -1355,32 +1339,69 @@ static X509 * php_openssl_x509_from_zval(zval * val, int makeresource, zend_reso
 		return NULL;
 	}
 
-	if (makeresource && resourceval) {
-		*resourceval = zend_register_resource(cert, le_x509);
-	}
 	return cert;
 }
 
+/* {{{ php_openssl_x509_from_param
+	Given a parameter, extract it into an X509 object.
+	The parameter can be:
+		. X509 object created using openssl_read_x509()
+		. if it starts with file:// then it will be interpreted as the path to that cert
+		. it will be interpreted as the cert data
+*/
+static X509 *php_openssl_x509_from_param(zend_object *cert_obj, zend_string *cert_str) {
+	if (cert_obj) {
+		return x509_certificate_from_obj(cert_obj)->x509;
+	}
+
+	ZEND_ASSERT(cert_str);
+
+	return php_openssl_x509_from_str(cert_str);
+}
+/* }}} */
+
+static X509 *php_openssl_x509_from_zval(zval *val, x509_certificate_object **cert_object)
+{
+	if (Z_TYPE_P(val) == IS_OBJECT && Z_OBJCE_P(val) == x509_certificate_ce) {
+		return x509_certificate_from_obj(Z_OBJ_P(val))->x509;
+	}
+
+	if (Z_TYPE_P(val) != IS_STRING && Z_TYPE_P(val) != IS_OBJECT) {
+		return NULL;
+	}
+
+	if (!try_convert_to_string(val)) {
+		return NULL;
+	}
+
+	return php_openssl_x509_from_str(Z_STR_P(val));
+}
 /* }}} */
 
 /* {{{ Exports a CERT to file or a var */
 PHP_FUNCTION(openssl_x509_export_to_file)
 {
-	X509 * cert;
-	zval * zcert;
+	X509 *cert;
+	zend_object *cert_obj;
+	zend_string *cert_str;
+
 	zend_bool notext = 1;
 	BIO * bio_out;
 	char * filename;
 	size_t filename_len;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "zp|b", &zcert, &filename, &filename_len, &notext) == FAILURE) {
-		RETURN_THROWS();
-	}
+	ZEND_PARSE_PARAMETERS_START(2, 3)
+		Z_PARAM_STR_OR_OBJ_OF_CLASS(cert_str, cert_obj, x509_certificate_ce)
+		Z_PARAM_PATH(filename, filename_len)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_BOOL(notext)
+	ZEND_PARSE_PARAMETERS_END();
+
 	RETVAL_FALSE;
 
-	cert = php_openssl_x509_from_zval(zcert, 0, NULL);
+	cert = php_openssl_x509_from_param(cert_obj, cert_str);
 	if (cert == NULL) {
-		php_error_docref(NULL, E_WARNING, "Cannot get cert from parameter 1");
+		php_error_docref(NULL, E_WARNING, "X.509 Certificate cannot be retrieved");
```

Comment:
I wasn't sure if this should be promoted to an exception or not. I think I'll try to improve the message to something like `Argument ... must be a valid X.509 certificate` after the other changes are ok.

Reply:
I think if we agree this should be a ValueError you could do it within the same PR (maybe as a separate commit?)

Reply:
I'd prefer different as it's not related to object conversion. It would apply to PEM file as well for example.